src/Symfony/Bundle/TwigBundle/TwigEngine.php

```diff
@@ -74,10 +74,15 @@ public function render($name, array $parameters = array())
             if ($name instanceof TemplateReference) {
                 try {
                     // try to get the real name of the template where the error occurred
-                    $name = $e->getTemplateName();
-                    $path = (string) $this->locator->locate($this->parser->parse($name));
+                    if (method_exists($e, 'getSourceContext')) {
+                        $templateName = $e->getSourceContext()->getName();
+                        $path = $e->getSourceContext()->getPath();
+                    } else {
+                        $templateName = $e->getTemplateName();
+                        $path = (string) $this->locator->locate($this->parser->parse($templateName));
+                    }
                     if (method_exists($e, 'setSourceContext')) {
-                        $e->setSourceContext(new \Twig_Source('', $name, $path));
+                        $e->setSourceContext(new \Twig_Source('', $templateName, $path));
```

Comment:
why not using ``$e->getSourceContext()`` directly here, instead of creating a new one with the same info ? Existence of the getter and setter go together (they have been introduced together)